Use std::span more in WebSocket code

cdumez opened the pull request:
#### de69356648cfb170e76b2e42d567049156356a71
<pre>
Use std::span more in WebSocket code
<a href="https://bugs.webkit.org/show_bug.cgi?id=271035">https://bugs.webkit.org/show_bug.cgi?id=271035</a>

Reviewed by Brent Fulgham.

Use std::span more in WebSocket code. This is work towards more widespread
std::span adoption in WebKit, which has security benefits.

* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WTF/wtf/StreamBuffer.h:
(WTF::StreamBuffer::append):
(WTF::StreamBuffer::firstBlockSpan const):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::fromUTF8):
* Source/WebCore/Modules/websockets/WebSocket.cpp:
(WebCore::WebSocket::didReceiveMessage):
(WebCore::WebSocket::didReceiveBinaryData):
* Source/WebCore/Modules/websockets/WebSocketChannelInspector.cpp:
(WebCore::WebSocketChannelInspector::createFrame):
* Source/WebCore/Modules/websockets/WebSocketChannelInspector.h:
* Source/WebCore/Modules/websockets/WebSocketDeflateFramer.cpp:
(WebCore::WebSocketDeflateFramer::deflate):
(WebCore::WebSocketDeflateFramer::inflate):
* Source/WebCore/Modules/websockets/WebSocketDeflater.cpp:
(WebCore::setStreamParameter):
(WebCore::WebSocketDeflater::addBytes):
(WebCore::WebSocketDeflater::finish):
(WebCore::WebSocketInflater::addBytes):
(WebCore::WebSocketInflater::finish):
* Source/WebCore/Modules/websockets/WebSocketDeflater.h:
(WebCore::WebSocketDeflater::bytes const):
(WebCore::WebSocketInflater::data const):
(WebCore::WebSocketInflater::bytes const):
(WebCore::WebSocketInflater::data): Deleted.
* Source/WebCore/Modules/websockets/WebSocketFrame.cpp:
(WebCore::WebSocketFrame::parseFrame):
(WebCore::appendFramePayload):
(WebCore::WebSocketFrame::makeFrameData):
(WebCore::WebSocketFrame::WebSocketFrame):
* Source/WebCore/Modules/websockets/WebSocketFrame.h:
(WebCore::WebSocketFrame::WebSocketFrame):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::Inspector::buildWebSocketMessage):
* Source/WebKit/WebProcess/Network/WebSocketChannel.cpp:
(WebKit::WebSocketChannel::notifySendFrame):
(WebKit::WebSocketChannel::createMessageQueue):
* Source/WebKit/WebProcess/Network/WebSocketChannel.h:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandle.cpp:
(WebCore::SocketStreamHandle::sendData):
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandle.h:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImpl.cpp:
(WebCore::SocketStreamHandleImpl::platformSend):
(WebCore::SocketStreamHandleImpl::platformSendHandshake):
(WebCore::SocketStreamHandleImpl::sendPendingData):
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImpl.h:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImplCFNet.cpp:
(WebCore::SocketStreamHandleImpl::platformSendInternal):
* Source/WebKitLegacy/WebCoreSupport/WebSocketChannel.cpp:
(WebCore::WebSocketChannel::send):
(WebCore::WebSocketChannel::startClosingHandshake):
(WebCore::WebSocketChannel::processFrame):
(WebCore::WebSocketChannel::enqueueRawFrame):
(WebCore::WebSocketChannel::processOutgoingFrameQueue):
(WebCore::WebSocketChannel::sendFrame):
* Source/WebKitLegacy/WebCoreSupport/WebSocketChannel.h:

Canonical link: <a href="https://commits.webkit.org/276224@main">https://commits.webkit.org/276224@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4cf636e1cea74965d4cb89b5de8c983dabd4e88f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46728 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/40113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46390 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20546 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44663 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20174 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/37962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/17357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/17649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/39062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2135 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/37411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/40244 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/39351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48312 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/43647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19063 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15628 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/20429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/41908 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/20649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50724 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6041 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/20056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10237 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -271,12 +271,14 @@ class String final {
     // Tries to convert the passed in string to UTF-8, but will fall back to Latin-1 if the string is not valid UTF-8.
     WTF_EXPORT_PRIVATE static String fromUTF8WithLatin1Fallback(const LChar*, size_t);
     static String fromUTF8WithLatin1Fallback(const char* characters, size_t length) { return fromUTF8WithLatin1Fallback(reinterpret_cast<const LChar*>(characters), length); }
+    static String fromUTF8WithLatin1Fallback(std::span<const LChar> characters) { return fromUTF8WithLatin1Fallback(characters.data(), characters.size()); }
```
This adaptor for `std::span<>` makes sense, but it's sad that `fromUTF8WithLatin1Fallback()` isn't using `span` natively; perhaps that could be a future change. (I.e., this change allows us to use `span` in more places, but doesn't improve the `fromUTF8WithLatin1Fallback()` code.)

cdumez replied:
Yes, that was me trying to keep the patch small. It's not a big change though, I'll update the default implementation to use span.

brentfulgham commented on Source/WebCore/Modules/websockets/WebSocketDeflateFramer.cpp:
```diff
@@ -189,13 +188,12 @@ std::unique_ptr<InflateResultHolder> WebSocketDeflateFramer::inflate(WebSocketFr
     frame.compress = false;
-    frame.payload = m_inflater->data();
-    frame.payloadLength = m_inflater->size();
+    frame.payload = m_inflater->bytes();
```
It's such an obvious thing, but reducing the amount of manual bookkeeping for these lengths is a really nice change.

cdumez replied:
That's one benefit. Another benefit is that std::span does bounds checking and provides functionalities such as `subspan()` which avoid pointer arithmetics and thus a lot of (security) bugs.

brentfulgham commented on Source/WebCore/Modules/websockets/WebSocketDeflater.cpp:
```diff
@@ -180,18 +180,17 @@ bool WebSocketInflater::addBytes(const uint8_t* data, size_t length)
     // Appends 4 octests of 0x00 0x00 0xff 0xff
     size_t consumedSoFar = 0;
-    while (consumedSoFar < strippedLength) {
+    while (consumedSoFar < strippedFields.size()) {
```
Isn't `strippedFields` size known at compile time? Maybe the compiler is smart enough to make `strippedFields.size()` a constant, but I wonder if this could be a `sizeof()` or something?

cdumez replied:
No need to try and optimize this. std::array::size() is constexpr (https://en.cppreference.com/w/cpp/container/array/size) so this does get resolved at compile time.

brentfulgham commented on Source/WebCore/Modules/websockets/WebSocketDeflater.cpp:
```diff
@@ -200,8 +199,8 @@ bool WebSocketInflater::finish()
         size_t availableCapacity = m_buffer.size() - writePosition;
-        size_t remainingLength = strippedLength - consumedSoFar;
-        setStreamParameter(m_stream.get(), strippedFields + consumedSoFar, remainingLength, m_buffer.data() + writePosition, availableCapacity);
+        size_t remainingLength = strippedFields.size() - consumedSoFar;
```
Again, we are performing a method call to get the constant value `4`, to do this calculation. But maybe this is optimized away.

brentfulgham commented on Source/WebCore/Modules/websockets/WebSocketFrame.cpp:
```diff
@@ -169,15 +168,14 @@ void WebSocketFrame::makeFrameData(Vector<uint8_t>& frameData)
     , reserved2(false)
     , reserved3(false)
     , masked(masked)
     , payload(payload)
```
I keep thinking of `std::span<>` like an object, and wanting you to `WTFMove()` it.

brentfulgham commented on Source/WebKit/NetworkProcess/curl/WebSocketTaskCurl.cpp:
```diff
@@ -345,15 +345,15 @@ std::optional<String> WebSocketTask::receiveFrames(Function<void(WebCore::WebSoc
             }
         } else
-            callback(frame.opCode, frame.payload, frame.payloadLength);
+            callback(frame.opCode, frame.payload.data(), frame.payload.size());
```
It would be nice if the `callback` could be modified to expect a `std::span<>`.

cdumez replied:
I agree but since this is Curl code I couldn't build locally, I mostly focused on keeping things building.

cdumez replied:
Anyway, you've convinced me. I'll fix the callback :)

brentfulgham commented on Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImpl.h:
```diff
@@ -51,17 +51,17 @@ class SocketStreamHandleImpl : public SocketStreamHandle {
-    WEBCORE_EXPORT void platformSendHandshake(const uint8_t* data, size_t length, const std::optional<CookieRequestHeaderFieldProxy>&, Function<void(bool, bool)>&&) final;
-    WEBCORE_EXPORT void platformClose() final;
+    void platformSend(std::span<const uint8_t> data, Function<void(bool)>&&) final;
+    void platformSendHandshake(const uint8_t* data, size_t length, const std::optional<CookieRequestHeaderFieldProxy>&, Function<void(bool, bool)>&&) final;
```
Why doesn't `platformSendHandshake` take a `span` now?

cdumez replied:
It should. I'm just trying to keep patch size reasonable so I have to stop "somewhere" :)

brentfulgham reviewed (approved):
This is such a nice cleanup! Looks like there might be more opportunities to do even more span work in follow-up patches! r=me

cdumez commented on Source/WebKit/NetworkProcess/cocoa/WebSocketTaskCocoa.mm:
```diff
@@ -91,7 +91,7 @@
             m_channel.didReceiveText(message.string);
         else
-            m_channel.didReceiveBinaryData(static_cast<const uint8_t*>(message.data.bytes), message.data.length);
+            m_channel.didReceiveBinaryData(std::span { static_cast<const uint8_t*>(message.data.bytes), message.data.length });
```
I'll try and land my other PR first so I can update this PR to call `toSpan(message.data)` here.

cdumez committed:
Use std::span more in WebSocket code
https://bugs.webkit.org/show_bug.cgi?id=271035

Reviewed by Brent Fulgham.

Use std::span more in WebSocket code. This is work towards more widespread
std::span adoption in WebKit, which has security benefits.

* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WTF/wtf/StreamBuffer.h:
(WTF::StreamBuffer::append):
(WTF::StreamBuffer::firstBlockSpan const):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::fromUTF8):
* Source/WebCore/Modules/websockets/WebSocket.cpp:
(WebCore::WebSocket::didReceiveMessage):
(WebCore::WebSocket::didReceiveBinaryData):
* Source/WebCore/Modules/websockets/WebSocketChannelInspector.cpp:
(WebCore::WebSocketChannelInspector::createFrame):
* Source/WebCore/Modules/websockets/WebSocketChannelInspector.h:
* Source/WebCore/Modules/websockets/WebSocketDeflateFramer.cpp:
(WebCore::WebSocketDeflateFramer::deflate):
(WebCore::WebSocketDeflateFramer::inflate):
* Source/WebCore/Modules/websockets/WebSocketDeflater.cpp:
(WebCore::setStreamParameter):
(WebCore::WebSocketDeflater::addBytes):
(WebCore::WebSocketDeflater::finish):
(WebCore::WebSocketInflater::addBytes):
(WebCore::WebSocketInflater::finish):
* Source/WebCore/Modules/websockets/WebSocketDeflater.h:
(WebCore::WebSocketDeflater::bytes const):
(WebCore::WebSocketInflater::data const):
(WebCore::WebSocketInflater::bytes const):
(WebCore::WebSocketInflater::data): Deleted.
* Source/WebCore/Modules/websockets/WebSocketFrame.cpp:
(WebCore::WebSocketFrame::parseFrame):
(WebCore::appendFramePayload):
(WebCore::WebSocketFrame::makeFrameData):
(WebCore::WebSocketFrame::WebSocketFrame):
* Source/WebCore/Modules/websockets/WebSocketFrame.h:
(WebCore::WebSocketFrame::WebSocketFrame):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::Inspector::buildWebSocketMessage):
* Source/WebKit/WebProcess/Network/WebSocketChannel.cpp:
(WebKit::WebSocketChannel::notifySendFrame):
(WebKit::WebSocketChannel::createMessageQueue):
* Source/WebKit/WebProcess/Network/WebSocketChannel.h:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandle.cpp:
(WebCore::SocketStreamHandle::sendData):
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandle.h:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImpl.cpp:
(WebCore::SocketStreamHandleImpl::platformSend):
(WebCore::SocketStreamHandleImpl::platformSendHandshake):
(WebCore::SocketStreamHandleImpl::sendPendingData):
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImpl.h:
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImplCFNet.cpp:
(WebCore::SocketStreamHandleImpl::platformSendInternal):
* Source/WebKitLegacy/WebCoreSupport/WebSocketChannel.cpp:
(WebCore::WebSocketChannel::send):
(WebCore::WebSocketChannel::startClosingHandshake):
(WebCore::WebSocketChannel::processFrame):
(WebCore::WebSocketChannel::enqueueRawFrame):
(WebCore::WebSocketChannel::processOutgoingFrameQueue):
(WebCore::WebSocketChannel::sendFrame):
* Source/WebKitLegacy/WebCoreSupport/WebSocketChannel.h:

Canonical link: https://commits.webkit.org/276224@main
